pkg/remote/client.go

```diff
@@ -0,0 +1,109 @@
+package remote
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	prompb "go.buf.build/grpc/go/prometheus/prometheus"
+)
+
+type HTTPConfig struct {
+	Timeout   time.Duration
+	TLSConfig *tls.Config
+	BasicAuth *BasicAuth
+	Headers   http.Header
+}
+
+type BasicAuth struct {
+	User, Password string
+}
+
+// WriteClient is a client implementation of the Prometheus remote write protocol.
+// It follows the specs defined by the official design document:
+// https://docs.google.com/document/d/1LPhVRSFkGNSuU1fBd81ulhsCPR4hkSZyyBj1SZ8fWOM
+type WriteClient struct {
+	hc  *http.Client
+	url *url.URL
+	cfg *HTTPConfig
+}
+
+func NewWriteClient(endpoint string, cfg *HTTPConfig) (*WriteClient, error) {
+	if cfg == nil {
+		cfg = &HTTPConfig{}
+	}
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	wc := &WriteClient{
+		hc: &http.Client{
+			Timeout: cfg.Timeout,
+		},
+		url: u,
+		cfg: cfg,
+	}
+	if cfg.TLSConfig != nil {
+		wc.hc.Transport = &http.Transport{
+			TLSClientConfig: cfg.TLSConfig,
+		}
+	}
+	return wc, nil
+}
+
+// Store sends a batch of samples to the HTTP endpoint,
+// the request is the proto marshaled and encoded.
+func (c *WriteClient) Store(ctx context.Context, series []*prompb.TimeSeries) error {
+	b, err := c.requestBody(series)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequestWithContext(
+		ctx, http.MethodPost, c.url.String(), bytes.NewReader(b))
+	if err != nil {
+		return fmt.Errorf("create new HTTP request failed: %w", err)
+	}
+	if c.cfg.BasicAuth != nil {
+		req.SetBasicAuth(c.cfg.BasicAuth.User, c.cfg.BasicAuth.Password)
+	}
+
+	req.Header.Set("User-Agent", "k6-prometheus-rw-output")
+
+	// They are mostly defined by the specs
+	req.Header.Add("Content-Encoding", "snappy")
+	req.Header.Set("Content-Type", "application/x-protobuf")
+	req.Header.Set("X-Prometheus-Remote-Write-Version", "0.1.0")
+
+	resp, err := c.hc.Do(req)
+	if err != nil {
+		return fmt.Errorf("HTTP POST request failed: %w", err)
+	}
+	defer resp.Body.Close()
+	io.Copy(io.Discard, resp.Body) //nolint:errcheck
+
+	if resp.StatusCode != http.StatusNoContent {
+		return fmt.Errorf("got status code: %s instead expected: 204 No Content", resp.Status)
+	}
+	return nil
+}
+
+func (c *WriteClient) requestBody(series []*prompb.TimeSeries) ([]byte, error) {
```

Comment:
Let's maybe detach it from the `c *WriteClient`; as long as it can be a simple function, it's better to keep it as it is.

Reply:
I'm not sure this is a good idea. The function is returning a `prompb.WriteRequest` why should we use it outside of the `WriteClient` context? For example, if we detach then we have to rename with something like `writeRequestEncodedBody`.

Reply:
Because at least now, it requires nothing from the actual `*WriteClient`. Even if the function returns encoded `prompb.WriteRequest` that doesn't mean that it should be in responsibility of the `WriteClient`. Simply it shouldn't be responsible for everything related to the writing :shrug: 

>For example, if we detach then we have to rename with something like `writeRequestEncodedBody`.

I see your point, but I can't agree that it sounds to everybody that if a `requestBody` method belongs to the `WriteClient` automatically helps to understand what it is without knowing the exact place where we use it or looking into its body. Another word I can understand is that it actually returns an encoded `prompb.WriteRequest` I either can get by a reading of the method or only together with the invocation place:
```
func (c *WriteClient) Store(ctx context.Context, series []*prompb.TimeSeries) error {
	b, err := c.requestBody(series)
	if err != nil {
		return err
	}
        ...
```

`writeRequestEncodedBody` sounds good to me, to be honest, but I'm fine even with the existing method name as long as it's the single request that the client is doing. Maybe just to reserve some place for the future and follow go conventions name it in a constructor like style `newWriteRequestBody()`.

Reply:
Changed and added a test for it